src/ml/neural_net/style_transfer/mps_style_transfer.m

```diff
@@ -0,0 +1,419 @@
+                                             contentLossNode.resultImage]];
+
+    totalLoss.resultImage.handle = [[TCMPSGraphNodeHandle alloc] initWithLabel:@"totalLossValue"];
+    totalLoss.resultImage.exportFromGraph = YES;
+    totalLoss.resultImage.synchronizeResource = YES;
+    totalLoss.resultImage.imageAllocator = [MPSImage defaultAllocator];
+
+    MPSNNInitialGradientNode *initialGradient
+      = [MPSNNInitialGradientNode nodeWithSource:totalLoss.resultImage];
+
+    BOOL resultsAreNeeded[] = { YES, YES };
+
+    NSArray<MPSNNFilterNode*>* lastNodes 
+      = [initialGradient trainingGraphWithSourceGradient:initialGradient.resultImage
+                                             nodeHandler: nil];
+
+    _trainingGraph = [MPSNNGraph graphWithDevice:_dev
+                                    resultImages:@[lastNodes[0].resultImage, lastNodes[1].resultImage]
+                                resultsAreNeeded:&resultsAreNeeded[0]];
+
+    _inferenceGraph = [MPSNNGraph graphWithDevice:_dev
+                                      resultImage:_model.forwardPass
+                              resultImageIsNeeded:YES];
+
+    _trainingGraph.format = MPSImageFeatureChannelFormatFloat32;
+    _inferenceGraph.format = MPSImageFeatureChannelFormatFloat32;
+  }
+  return self;
+}
+ 
+- (NSDictionary<NSString *, NSData *> *)exportWeights {
+  return [_model exportWeights];
+}
+
+- (NSDictionary<NSString *, NSData *> *)predict:(NSDictionary<NSString *, NSData *> *)inputs {
+  NSUInteger contentWidth;
+  NSUInteger contentHeight;
+  NSUInteger contentChannels;
+
+  [inputs[@"imageWidth"] getBytes:&contentWidth length:sizeof(contentWidth)];
+  [inputs[@"imageHeight"] getBytes:&contentHeight length:sizeof(contentHeight)];
+  [inputs[@"imageChannels"] getBytes:&contentChannels length:sizeof(contentChannels)];
+
+  MPSImageDescriptor *imgDesc = [MPSImageDescriptor
+    imageDescriptorWithChannelFormat:MPSImageFeatureChannelFormatFloat32
+                               width:contentWidth
+                              height:contentHeight
+                     featureChannels:contentChannels
+                      numberOfImages:1
+                               usage:MTLTextureUsageShaderWrite | MTLTextureUsageShaderRead];
+
+  NSMutableArray<MPSImage *> *contentImageArray;
+
+  for (NSUInteger index = 0; index < _batchSize; index++) {
+    NSString* key = [NSString stringWithFormat:@"%@%lu", @"contentImage", index];
+    MPSImage *contentImage = [[MPSImage alloc] initWithDevice:_dev imageDescriptor:imgDesc];
+    [contentImage writeBytes:inputs[key].bytes dataLayout:(MPSDataLayoutHeightxWidthxFeatureChannels)imageIndex:0];
+    [contentImageArray addObject:contentImage];
+  }
+
+  MPSImageBatch *contentImageBatch = [contentImageArray copy];  
+
+  id<MTLCommandBuffer> cb = [_commandQueue commandBuffer];
+
+  NSMutableArray *intermediateImages = [[NSMutableArray alloc] initWithArray:@[]];
```

Comment:
The RHS here can just be `[[NSMutableArray alloc] init]`

Reply:
(Some people prefer `[NSMutableArray new]` which in theory is equivalent and simpler, but I'm personally just not used to it)

Reply:
Consistently allocated and initiated NS* stuff.